[Docs] Add info on logs on Kubernetes & metrics in the ops section (#10525)
digital-asset/daml: docs/source/ops/connect/content.rst

.. Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _ops-connect-content_index:

> Log aggregation on Kubernetes

Reviewer: ```suggestion
Log Aggregation On Kubernetes
```

=============================

> If configured so, every daml service can output its logs in JSON format.

Reviewer: ```suggestion
If configured so, every daml service can output its logs in JSON format on stdout.
```

This makes it easy to analyze these with search engines like `ElasticSearch <https://www.elastic.co/elasticsearch/>`_ in combination with `Kibana <https://www.elastic.co/kibana/>`_.

To collect all logs for analyzing them e.g. in your `ElasticSearch` instance, one easy way to achieve this is using `Filebeat <https://www.elastic.co/beats/filebeat>`_.
It regulary collects the logs of your containers and ingests them into `Elasticsearch`, `Logstash <https://www.elastic.co/logstash/>`_ or `Kafka <https://kafka.apache.org/>`_ etc. and the best is, it doesn't require any additional configuration for the daml services.

> Daml Metrics options

Reviewer: ```suggestion
Daml Metrics Options
```

====================

Most daml services support the option to export metrics for application insight.
Currently we support exporting to these with the corresponding cli option:

- `Prometheus <https://prometheus.io>`_ via ``--metrics-reporter "prometheus://HOST[:PORT]"``

- `Graphite <https://graphiteapp.org>`_ via ``--metrics-reporter "graphite://HOST[:PORT][/METRIC_PREFIX]"``

- console output (logs) via ``--metrics-reporter console``

> Deployment example with minikube

Reviewer: ```suggestion
Deployment Example With Minikube
```

================================

Prerequisites
--------------

> For this to work, you need to have `docker` & `minikube` installed (If you run into any issue with the kubernetes proxy and minikube on Ubuntu, this is the fix [https://github.com/kubernetes-sigs/kind/issues/2240#issuecomment-888213835]).

Reviewer: Does that work as a link? I've never seen that syntax in RST documentation.

Reviewer: Maybe:
```suggestion
For this to work, you need to have `docker` & `minikube` installed (If you run into any issue with the kubernetes proxy and minikube on Ubuntu, `this fix may be worth trying <https://github.com/kubernetes-sigs/kind/issues/2240#issuecomment-888213835>`_).
```

> Installing elasticsearch on your local kubernetes cluster

Reviewer: ```suggestion
Installing Elasticsearch On Your Local Kubernetes Cluster
```

=========================================================

> First you need to add the elastic repository to helm, so the chart can be found.

Reviewer: ```suggestion
First you need to add the Elastic repository to helm, so the chart can be found.
```

``
helm repo add elastic https://helm.elastic.co
helm repo update
``

> Now you can setup the elastic search cluster via:

Reviewer: ```suggestion
Now you can setup the Elasticsearch cluster via:
```

``
helm upgrade --install es --version 7.9.3 elastic/elasticsearch --values https://raw.githubusercontent.com/elastic/helm-charts/master/elasticsearch/examples/minikube/values.yaml
``

> Side note: Please do not use this config setup in production as antiAffinity is set to soft, which effectively allows elasticsearch to run all pods on one physical machine. However for local testing with minikube this is necessary and thus fine. 

Reviewer: ```suggestion
Side note: Please do not use this config setup in production as ``antiAffinity`` is set to ``soft``, which effectively allows Elasticsearch to run all pods on one physical machine. However for local testing with minikube this is necessary and thus fine. 
```

.. Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _ops-connect-content_index:

Log aggregation on Kubernetes
=============================

If configured so, every daml service can output it's logs in JSON format.
This makes it easy to analyze these with search engines like `ElasticSearch <https://www.elastic.co/elasticsearch/>`_ in combination with `Kibana <https://www.elastic.co/kibana/>`_.

To collect all logs for analyzing them e.g. in your `ElasticSearch` instance, one easy way to achieve this is using `Filebeat <https://www.elastic.co/beats/filebeat>`_.
It regulary collects the logs of your containers and ingests them into `Elasticsearch`, `Logstash <https://www.elastic.co/logstash/>`_ or `Kafka <https://kafka.apache.org/>`_ etc. and the best is, it doesn't require any additional configuration for the daml services.

> You can find external documentation on, how to setup `ElasticSearch` with `Filebeat` and `Kibana` for analyzing logs on your Kubernetes cluster `here <https://www.deepnetwork.com/blog/2020/01/27/ELK-stack-filebeat-k8s-deployment.html>`_.

Reviewer 2: We should turn this into a full section that shows a fully working setup and doesn’t just link to an external blogpost.

Author: I can do that, however to what extend do we want to do it here?
E.g. my first thought would be to make it like a blog post where one shows how to create that setup on a local minikube cluster.

Reviewer 2: Basing it on a minicube example sounds sensible :+1:

> Daml Metrics options

Reviewer 2: Similar here, this should show an example of integrating this with prometheus and not just list CLI flags.

Author: Same as for the other, Minikube cluster setup?

Author: Side note, I would like to keep these sections and add an additional example section because I think it is good to have an overview + example.

Author: Great, many thanks for the input. Then I can proceed further here :)
